presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -447,15 +452,31 @@ private OuterJoinPushDownResult processLimitedOuterJoin(Expression inheritedPred
                 }
             }
 
-            // See if we can push down any outer or join predicates to the inner side
-            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(and(outerEffectivePredicate, joinPredicate))) {
+            // See if we can push down any outer predicates to the inner side
+            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(outerEffectivePredicate)) {
                 Expression rewritten = potentialNullSymbolInference.rewriteExpression(conjunct, not(in(outerSymbols)));
                 if (rewritten != null) {
                     innerPushdownConjuncts.add(rewritten);
                 }
             }
 
-            // TODO: consider adding join predicate optimizations to outer joins
+            // See if we can push down any join predicates to the inner side
+            for (Expression conjunct : extractConjuncts(joinPredicate)) {
+                Expression rewritten = innerInference.rewriteExpression(conjunct, not(in(outerSymbols)));
+                if (rewritten != null) {
+                    innerPushdownConjuncts.add(rewritten);
+                    if (!shouldRemoveOuterJoinConjunct(outerSymbols, conjunct)) {
```

Comment:
So I took a crack at writing the method to see how it would look and this is what I got. After doing so, I think there are 3 main things I want to point out:
-1. We need to be stripping out the non-deterministic join predicate conjuncts from being considered for push down
```java
            joinConjuncts.addAll(filter(extractConjuncts(joinPredicate), not(DeterminismEvaluator::isDeterministic)));
            joinPredicate = stripNonDeterministicConjuncts(joinPredicate);
```
-2. potentialNullSymbolInference is valid and better for rewriting the non-inferrable conjuncts of joinPredicate
-3. We can use the existing equality inference  partitioning behavior to generate which parts of the join predicate equalities to push down:
```java
            EqualityInference.EqualityPartition joinEqualityPartition = createEqualityInference(joinPredicate).generateEqualitiesPartitionedBy(not(in(outerSymbols)));
            innerPushdownConjuncts.addAll(joinEqualityPartition.getScopeEqualities());
            joinConjuncts.addAll(joinEqualityPartition.getScopeComplementEqualities())
                    .addAll(joinEqualityPartition.getScopeStraddlingEqualities());
```
It should be similar to what you were doing before, but without introducing much new code.